pallets/automation-time/src/migrations.rs

```diff
@@ -320,3 +322,198 @@
 // 		}
 // 	}
 // }
+
+pub mod v4 {
+	use super::*;
+	use crate::{AccountTaskId, ScheduledTasks};
+	use frame_support::{migration::storage_key_iter, BoundedVec, Twox64Concat};
+
+	pub const OLD_STORAGE_PREFIX: &[u8] = b"ScheduledTasksV2";
+	pub const NEW_STORAGE_PREFIX: &[u8] = b"ScheduledTasksV3";
+
+	// Use weight to determine when scheduled task slot is full
+	pub struct MigrateToV4<T>(PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV4<T> {
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<(), &'static str> {
+			use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+			// Get count per scheduled time
+			let mut pre_scheduled_count: Vec<(UnixTime, u32)> = vec![];
+			storage_key_iter::<u64, BoundedVec<AccountTaskId<T>, ConstU32<256>>, Twox64Concat>(
+				PALLET_PREFIX,
+				OLD_STORAGE_PREFIX,
+			)
+			.for_each(|(time, task_ids)| {
+				pre_scheduled_count.push((time, task_ids.len() as u32));
+			});
+			Self::set_temp_storage::<Vec<(UnixTime, u32)>>(
+				pre_scheduled_count,
+				"pre_scheduled_count",
+			);
+
+			log::info!(
+				target: "automation-time",
+				"migration: AutomationTime storage version v4 PRE migration checks succesful!"
+			);
+
+			Ok(())
+		}
+
+		fn on_runtime_upgrade() -> Weight {
+			log::info!(target: "automation-time", "Migrating automation-time v4");
+
+			// Move all tasks from ScheduledTasksV2 to ScheduledTasksV3
+			let mut task_count = 0u64;
+			let mut migrated_keys_count = 0u64;
+			storage_key_iter::<UnixTime, BoundedVec<AccountTaskId<T>, ConstU32<256>>, Twox64Concat>(
+				PALLET_PREFIX,
+				OLD_STORAGE_PREFIX,
+			)
+			.drain()
+			.for_each(|(time, task_ids)| {
+				migrated_keys_count += 1;
+				let weight = task_ids.clone()
+					.into_iter()
+					.fold(0u128, |acc, (account_id, task_id)| {
+						task_count += 1;
+						if let Some(task) = Pallet::<T>::get_account_task(account_id.clone(), task_id) {
+							acc + task.action.execution_weight::<T>() as u128
+						} else {
+							log::warn!(target: "automation-time", "Unable to get task with id {:?} for account {:?}", task_id, account_id);
+							acc
+						}
+					});
+
+				// Insert all tasks without checking MaxWeightPerSlot to allow existing tasks to be run
+				ScheduledTasksV3::<T>::insert(time, ScheduledTasks{ tasks: task_ids.to_vec(), weight });
+			});
+
+			// 1 read to load iterator
+			// task_count reads to load tasks
+			// For each time (migrated_key_count) in scheduled tasks there is
+			// 1 read to get it into memory
+			// 1 write to remove it from the old scheduled tasks
+			// 1 write to add it to the new scheduled tasks
+			let weight = T::DbWeight::get()
+				.reads_writes(migrated_keys_count + task_count + 1, migrated_keys_count * 2);
+
+			// Adding a buffer for the rest of the code
+			weight + 10_000_000 + (migrated_keys_count * 15_000_000)
```

Comment:
Running benchmarks to prove these numbers

Reply:
@rhuttman mind double checking if I did something stupid in the benchmarks? I ran them overnight on my laptop and made very little progress. Doesn't seem to be an infinite loop because it is progressing just soooooo slowly.

Reply:
Oh. I think it's because notify tasks are so much cheaper to run. Current weight limits are set for ~500 transfers but that's ~15,000 notifies. Thoughts on how I should handle this? The diff between the cost of a notify task and the auto compound task are so large that it makes it difficult to set this limit.

Reply:
Resolved this by re-adding MaxTasksPerSlot as a hard limit for light weight tasks